bitcoin/src/consensus/encode.rs

```diff
@@ -156,11 +180,14 @@ pub trait WriteExt: io::Write {
     /// Outputs an 8-bit signed integer.
     fn emit_i8(&mut self, v: i8) -> Result<(), io::Error>;
 
+    /// Outputs a variable sized integer.
+    fn emit_varint(&mut self, v: impl IntoU64) -> Result<usize, io::Error>;
```

Comment:
I'm a bit confused about what `impl IntoU64` is supposed to mean. I have never seen this used in an argument to a trait method before and this style isn't used anywhere else in this codebase.

It would be clearer to just use a generic. The result will still not be object-safe. To get object safety you'd need to use a `dyn` type.

Reply:
That use of `impl` is just a shorthand to create a generic, these two signatures are equivalent (except the ability of the caller to explicitly specify the generic type using the turbofish):
```rust
    fn emit_varint(&mut self, v: impl IntoU64) -> Result<usize, io::Error>;
```
```rust
    fn emit_varint<T: IntoU64>(&mut self, v: T) -> Result<usize, io::Error>;
```

I can add `Self: Sized`, but not sure that is sufficient to make the `WriteExt` types boxable. Let me try.

Reply:
Hmm, that's pretty annoying as then we don't allow reading from and writing to byte slices anymore as they are `?Sized`. Is the boxability really an issue? I would really prefer to not get a dynamic reference in this method.

Reply:
Thanks for explaining. And no, I don't think `Sized` is sufficient to make `WriteExt` boxable. A generic method is inherently not boxable because the vtable needs to hold pointers to actual functions, and the compiler doesn't know upfront how many different functions the generic will expand into.

I think we should sacrifice object-safety. But if we want to keep it, our options are:

* Return to the old code which has a type rather than methods
* Move the method out of the trait definition into a standalone method
* Move the method into `IntoU64` rather than `WriteExt` (and then `IntoU64` will not be object-safe but that seems less bad)

Reply:
- change the signature to take u64 and convert as call-time

But I agree object-safety here shouldn't be important.

Reply:
I don't see how it's related to slices, can you explain? This method previously wasn't present and the trait was object safe, now the method was added which makes it no longer object safe which is easily fixable by doing `where Self: Sized`, so that any existing code using `&mut dyn WriteExt` will continue to work without us losing anything.

Also I suspect we should `impl WriteExt for &mut dyn WriteExt` and `impl WriteExt for Box<dyn WriteExt>` but that's off topic.

Reply:
> easily fixable by doing `where Self: Sized`,

I don't understand this. The problem with object-safety is that we are adding a generic method to the trait. `Sized` won't help with this.

Reply:
It will, the offending method will be unavailable in unsized context but all others will be fine so any code using `dyn WriteExt` today will continue to work.

Reply:
>the offending method will be unavailable in unsized context

ooo TIL. Ok, sorry for causing confusion.